src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java

```diff
@@ -306,4 +374,123 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+    }
+
+    /**
+     * Get MultiTenantSite Class by Reflection 
+     * @return
+     * @throws Throwable 
+     */
+    protected Class<?> getMultiTenantClass() throws ClassNotFoundException, LinkageError  {
+        try {
+            Class<?> c = ClassUtils.forName(MULTITENANT_SITE_CLASSNAME, getClass().getClassLoader());
+            return c;
+        } catch (ClassNotFoundException | LinkageError e) {
+            throw e;
+
+        }
+    }
+
+
+    /**
+     * Gets getParentId method by reflection   
+     * @return
+     * @throws NoSuchMethodException, SecurityException,ClassNotFoundException, LinkageError 
+     */
+    protected Method getMultiTenantGetParentSiteIdMethod() throws NoSuchMethodException, SecurityException,ClassNotFoundException, LinkageError {
+        try {
+            return getMultiTenantClass().getMethod(MULTITENANTSITE_GETPARENTID_METHODNAME);
+        } catch (NoSuchMethodException | SecurityException | ClassNotFoundException | LinkageError  e) {
+            // TODO Auto-generated catch block
+            throw e;
+        }
+    }
+
+
+    /**
+     * Reflection getParentSiteId Invocation 
+     * @param site
+     * @return
+     */
+    protected Long invokeMultiTenantGetParentSiteIdMethod(Object site) {
+        Long result = 0L;
+        try {
+            String parentSiteId =ReflectionUtils.invokeMethod(getMultiTenantGetParentSiteIdMethod(), site).toString();
+            if (parentSiteId!=null) {
+                result =Long.parseLong(parentSiteId);
+            }        
+            return result;
+        } catch ( SecurityException | IllegalArgumentException | NoSuchMethodException | ClassNotFoundException | LinkageError e) {
+            LOG.error("Problem invoking getParentSiteId method from reflected Multitenant class", e);
+            return 0L;
+
+        }
+        
+    }
+
+
+    /**
+     * Get the parent from a MultiTentantSite.
+     * The MultiTenantSite environment is detected by reflection.
+     * @param resourceName
+     * @return
+     */
+    protected String getMultiTenantSiteSpecificResourceNameParent(String resourceName) {
+        BroadleafRequestContext brc = BroadleafRequestContext.getBroadleafRequestContext();
+        if (brc != null) {
+            //getMultiTenantClass retrieves the class by reflection, the casts.
+            Object site="";
+            try {
+                site = (getMultiTenantClass().cast(brc.getNonPersistentSite()));
+                String siteDirectory = getSiteDirectory((invokeMultiTenantGetParentSiteIdMethod(site)));
+                return FilenameUtils.concat(siteDirectory, resourceName);
+            } catch (ClassNotFoundException | LinkageError e) {
+                //The MultitenantEnvironment should be checked before reaching this point so
+                //this exception should never occur if the MultiTenant environment is ok
+               LOG.error("Problem reflecting Multitenant Class", e);
```

Comment:
You should not be erroring here. It is a perfectly valid case to not find the multitenant class to get the parent id, if multitenant is not on your classpath.

Reply:
Yes, as I'm asking before is the class exists with isMultiTenantEnvironment I should never get into this exception... but....
1. In a previous comment, Brian say we never bury any exception.

2. Let's suppose someone is extending this class and is somehow bypassing the isMultiTenantEnvironment call. In that case it makes sense to log the exception.

Reply:
Ok, I guess I see that now but it was pretty confusing for me to trace down.

As long as you have absolutely tested these changes in the DemoSite project (where no multitenancy exists at all) and you do not get any extraneous error logs, then we are good to go.

Reply:
In Demo site is working, I don't see errors. I worth to do a double check.
I used DemoSite solution as base, as it has no multitenant (and is reflected in S3 as well).
I don't know if Demo Site has a limitation I should consider.

Reply:
Cool, as long as you have validated that everything works fine in demo site with the S3 upload (sounds like you validated it on the S3 side) and you can load the image in a browser, then we are good to merge.

Reply:
@egorbatik Can you confirm that this is good to merge based on @phillipuniverse's  comment?

Reply:
Good to merge